Common/Orders/Fills/EquityFillModel.cs

```diff
@@ -587,11 +587,52 @@ public virtual OrderEvent MarketOnCloseFill(Security asset, MarketOnCloseOrder o
             {
                 return fill;
             }
+
+            var subscribedTypes = GetSubscribedTypes(asset);
+
+            // The official close is recorded after the market is closed
+            if (subscribedTypes.Contains(typeof(Tick)))
+            {
+                var primaryExchange = (byte)((Equity)asset).PrimaryExchange;
+                var officialClose = (uint)(TradeConditionFlags.Regular | TradeConditionFlags.OfficialClose);
+                var closingPrints = (uint)(TradeConditionFlags.Regular | TradeConditionFlags.ClosingPrints);
+                var ticks = asset.Cache.GetAll<Tick>().OrderBy(x => x.EndTime).ToList();
+
+                // Get the last valid (non-zero) tick of trade type from an close market
+                var trade = ticks
+                    .Where(x => !string.IsNullOrWhiteSpace(x.SaleCondition))
+                    .LastOrDefault(x =>
+                        x.TickType == TickType.Trade && x.Price > 0 && x.ExchangeCode == primaryExchange &&
+                        (x.ParsedSaleCondition == officialClose || x.ParsedSaleCondition == closingPrints));
+
+                // If there is no tick with the desired information, we wait one minute
+                if (trade == null)
+                {
+                    trade = ticks.LastOrDefault(x => x.Price > 0);
+                    if (trade == null || (trade.EndTime - nextMarketClose).TotalMinutes < 1)
+                    {
+                        return fill;
+                    }
+                }
+
+                fill.FillPrice = trade.Price;
+            }
             // make sure the exchange is open/normal market hours before filling
-            if (!IsExchangeOpen(asset, false)) return fill;
+            else if (!IsExchangeOpen(asset, false))
```

Comment:
> if (asset.LocalTime < nextMarketClose)

At line 586, means asset time would already be past market close of the order,

> if (!IsExchangeOpen(asset, false))

Requires the market to be open to skip this `if`?
Meaning we really are waiting for the next market open to fill the MOC order, for resolution != Tick?

Reply:
`IsExchangeOpen` tests if the market is open (same as 586), if not, check the bar size, this handles large-sized bars (hours/days) in order to fill at that time, not on the next open.